[Concurrency] Alternative Task API.

DougGregor opened the pull request:
The `Task` type has oscillated somewhat from being purely a namespace,
to having instances that are used (albeit rarely), back to purely
being a namespace that isn't used for all that many names. Many of the
names that used to be on Task have already been moved out, e.g., for
creating new detached tasks, creating new task groups, adding
cancellation handlers, etc.

Collapse `Task.Handle<Success, Failure>` into `Task<Success, Failure>`.
`Task.Handle` is the type that is most frequently referenced in the
concurrency library, so giving it the short name `Task` is most
appropriate. Replace the top-level async/detach functions with a
`Task` initializer and `Task.detached`, respectively.

The `Task` type can still act as a namespace for static operations
such as, e.g., `Task.isCancelled`. Do this with an extension of the
form:

    extension Task where Success == Never, Failure == Never { ... }

We've been accruing a number of compatibility shims. Move them all
into their own source file, deprecate them, and make them
always-emit-into-client so they don't have any ABI impact.

This also covers other changes that came up in the review:
* Switch `get()` to `value` as an async property
* Switch `getResult()` to `result` as an async property
* Switch order/names of `withTaskCancellationHandler` parameters
* Update names of `TaskPriority` members, and make it a `RawRepresentable` struct

DougGregor committed:
[Concurrency] Alternative Task API.

The `Task` type has oscillated somewhat from being purely a namespace,
to having instances that are used (albeit rarely), back to purely
being a namespace that isn't used for all that many names. Many of the
names that used to be on Task have already been moved out, e.g., for
creating new detached tasks, creating new task groups, adding
cancellation handlers, etc.

Collapse `Task.Handle<Success, Failure>` into `Task<Success, Failure>`.
`Task.Handle` is the type that is most frequently referenced in the
concurrency library, so giving it the short name `Task` is most
appropriate. Replace the top-level async/detach functions with a
`Task` initializer and `Task.detached`, respectively.

The `Task` type can still act as a namespace for static operations
such as, e.g., `Task.isCancelled`. Do this with an extension of the
form:

    extension Task where Success == Never, Failure == Never { ... }

We've been accruing a number of compatibility shims. Move them all
into their own source file, deprecate them, and make them
always-emit-into-client so they don't have any ABI impact.

DougGregor committed:
Move CancellationError out to the top level

DougGregor commented:
@swift-ci please test

DougGregor committed:
Fix up some documentation for the async/detach renamings

ktoso reviewed (approved):
Looking good! Thanks Doug, pretty happy with the final names of things.

Minor change to be done in priorities still, you can do this here or later.

DougGregor committed:
[Concurrency] Rename TaskPriority members.

Based on feedback from the second review, we decided to go with
high/default/low/background, with aliases for the Dispatch-inspired
names. While here, make TaskPriority be backed by a UInt8 to better
describe the actual restrictions, and start removing userInteractive,
because clients shouldn't be able to specify it.

DougGregor committed:
Update task cancellation API based on review feedback

Swap the order of the arguments so we now have:

    await withTaskCancellationHandler {
      // do stuff
    } onCancel: {
      // ..
    }

DougGregor committed:
[Concurrency] Add async properties Task.value and Task.result

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain
